python/nav/alertengine/base.py

```diff
@@ -249,177 +258,243 @@ def _check_permissions(account, alert, alertsubscription, dupemap, _logger,
 
 
 def handle_queued_alerts(queued_alerts, now=None):
-    """Handles profile-queued alerts for later dispatch"""
+    """Processes and dispatches notifications that are queued for individual users.
+    Every AccountAlertQueue object represents a specific AlertQueue item that has been
+    scheduled for dispatch to a particular user. Typical scheduling choices may be:
+
+      * Immediately.
+      * Wait until the next timeperiod in the user's profile becomes active.
+      * Wait until the daily dispatch time set in the profile.
+      * Wait until the weekly dispatch time set in the profile.
+
+    If a user's profile says so, a delayed notification may also be ignored and removed
+    from the queue if the represented alert has been resolved since it was added to the
+    queue.
+
+    :param queued_alerts: A sequence of all AccountAlertQueue items to consider.
+    :type queued_alerts: List[AccountAlertQueue]
+    :param now: Set only when testing, to force what value should be considered the
+                current time.
+    :type now: datetime
+
+    :return: A tuple of stats on how the queued alerts were processed:
+             (accounts_sent_daily, accounts_sent_weekly, num_sent_alerts,
+              num_failed_sends, num_resolved_alerts_ignored)
+    """
     _logger = logging.getLogger('nav.alertengine.handle_queued_alerts')
 
     if not now:
         now = datetime.now()
 
-    # We want to keep track of wether or not any weekly or daily messages have
-    # been sent so that we can update the state of the users
-    # last_sent_daily/weekly
-    sent_weekly = []
-    sent_daily = []
+    # We want to keep track of which users have received their weekly or daily
+    # notifications during this queue run, so that their profile states can be updated
+    accounts_sent_weekly = set()
+    accounts_sent_daily = set()
 
     num_sent_alerts = 0
     num_resolved_alerts_ignored = 0
     num_failed_sends = 0
 
-    for queued_alert in queued_alerts:
-        send, daily, weekly = False, False, False
+    for queued_alert in queued_alerts:  # type: AccountAlertQueue
+        result = process_single_queued_notification(queued_alert, now=now)
 
-        try:
-            subscription = queued_alert.subscription
-        except AlertSubscription.DoesNotExist:
-            _logger.error('account queued alert %d does not have subscription, '
-                          'probably a legacy table row', queued_alert.id)
-            continue
+        # The result value is processed purely for statistics and record keeping
+        if result == DISPATCH_FAILED:
+            num_failed_sends += 1
+        elif result == DISPATCH_IGNORED:
+            num_resolved_alerts_ignored += 1
+        elif result in (DISPATCHED_NOW, DISPATCHED_WEEKLY, DISPATCHED_DAILY):
```

Comment:
Can result be anything else? Would an `else:` be sufficient here?

Reply:
Maybe not at the moment, but these are currently the only result codes that definitely fall in the "a notification was sent" category. I thought it best to keep it like this in case someone forgot to update the client code when the implementation starts returning new values in the future ;-)